### PR TITLE
CI: Check Debug and Release builds with -Werror

### LINKED ISSUE #1759: CI: add a test for warnings turned errors
In both Debug and Release mode. Right now we are getting a lot of warning messages, especially in Release mode.

### PR DESCRIPTION
**Pull request by certik**
We already had a dedicated Debug build. A new Release build was added. This new build is similar to the "matrix" builds, which are also Release, but they use platform independent build using Xonsh, while this new build uses a Ubuntu specific build, analogous to the Debug build.

Fixes #1759.

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -4609,10 +4609,9 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                     case (ASR::Pow):
                         result = std::pow(left_value, right_value);
                         break;
-                    // Reconsider
                     default: {
                         LCOMPILERS_ASSERT(false);
-                        op = ASR::binopType::Pow;
+                        result = left_value; // silence a warning
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:4614]: Raise an exception here.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -4647,10 +4646,9 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:4651]: Raise an exception here.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 3]
```diff
@@ -787,6 +787,7 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
             der_type = ASR::down_cast<ASR::StructType_t>(der_sym);
         } else {
             LCOMPILERS_ASSERT(false);
+            return nullptr; // silence a warning
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:790]: Raise an exception here.

File: src/libasr/codegen/c_utils.h [hunk 4]
```diff
@@ -451,6 +451,7 @@ class CCPPDSUtils {
                 return get_tuple_type(tup_type);
             }
             LCOMPILERS_ASSERT(false);
+            return ""; // To silence a warning
```
Comment (certik) [inline src/libasr/codegen/c_utils.h:454]: Raise an exception here.

File: src/libasr/pass/instantiate_template.cpp [hunk 5]
```diff
@@ -373,7 +373,7 @@ class FunctionInstantiator : public ASR::BaseExprStmtDuplicator<FunctionInstanti
                 switch (op) {
                     case (ASR::binopType::Add): { result = left_value + right_value; break; }
                     case (ASR::binopType::Div): { result = left_value / right_value; break; }
-                    default: { LCOMPILERS_ASSERT(false); } // should never happen
+                    default: { LCOMPILERS_ASSERT(false); result=0; } // should never happen
```
Comment (certik) [inline src/libasr/pass/instantiate_template.cpp:376]: Raise an exception here.

File: src/libasr/pass/instantiate_template.cpp [hunk 6]
```diff
@@ -388,7 +388,7 @@ class FunctionInstantiator : public ASR::BaseExprStmtDuplicator<FunctionInstanti
                 switch (op) {
                     case (ASR::binopType::Add): { result = left_value + right_value; break; }
                     case (ASR::binopType::Div): { result = left_value / right_value; break; }
-                    default: { LCOMPILERS_ASSERT(false); }
+                    default: { LCOMPILERS_ASSERT(false); result = 0; }
```
Comment (certik) [inline src/libasr/pass/instantiate_template.cpp:391]: Raise an exception here.

File: src/libasr/pass/pass_compare.cpp [hunk 7]
```diff
@@ -149,6 +149,7 @@ class CompareExprReplacer : public ASR::BaseExprReplacer<CompareExprReplacer>
             }
             default: {
                 LCOMPILERS_ASSERT(false);
+                return nullptr; // silence a warning
```
Comment (certik) [inline src/libasr/pass/pass_compare.cpp:152]: Raise an exception here.

File: src/libasr/pass/pass_utils.cpp [hunk 8]
```diff
@@ -938,6 +938,8 @@ namespace LCompilers {
                         // cond: cond1 || cond2
                         cond = ASRUtils::EXPR(make_LogicalBinOp_t(al, loc,
                             cond1, ASR::logicalbinopType::Or, cond2, log_type, nullptr));
+                        // TODO: is cmp_op uninitialized here?
+                        cmp_op = ASR::cmpopType::LtE; // silence a warning
```
Comment (certik) [inline src/libasr/pass/pass_utils.cpp:942]: Raise an exception here.

**Comment by certik** [comment]
@Smit-create, @Shaikh-Ubaid if you want to finish this one, that would be awesome. 

TODO:

* [ ] First fix all the warnings at the CI in the Release mode, so that it builds
* [ ] Some of my "fixes" seem to introduce (or reveal) bugs, see the macOS failure introduced in this PR, we need to fix those.

After the above two points, I suggest to merge, and then we can fix the other issues revealed in this PR, see my comments above.

See also #1765 which happens even in master for me, not sure if it is related or not.

**Comment by Shaikh-Ubaid** [comment]
It seems there was only the `nbody` test case that was failing. I commented it out in [TEST: Comment out failing nbody test](https://github.com/lfortran/lfortran/pull/1764/commits/cabb3d3d222d2e27d74e8f11bce7edc80a29d68c).

**Comment by Shaikh-Ubaid** [comment]
Ready.

**Comment by Smit-create** [comment]
Thanks, @certik  and @Shaikh-Ubaid. I see that the CI is green now. Do we want to fix the `nbody` test in this PR or merge this and fix it in a new PR?

**Comment by Shaikh-Ubaid** [comment]
>Do we want to fix the nbody test in this PR or merge this and fix it in a new PR?

For `nbody` it seems we have a dedicated issue at https://github.com/lfortran/lfortran/issues/1765. Also, this PR seems to tackle the issue https://github.com/lfortran/lfortran/issues/1759. Thus, I assumed it is safer to comment out the failing test instead of fixing in this PR.

**Comment by certik** [comment]
Thanks @Shaikh-Ubaid! I think this looks good to merge now, so I'll merge it and we'll improve upon things later.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 9]
```diff
@@ -4652,8 +4653,7 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
         if( ASRUtils::is_data_only_array(array_type, ASR::abiType::Source) &&
             ASRUtils::expr_intent(array_section->m_v) != ASR::intentType::Local ) {
             ASR::dimension_t* m_dims = nullptr;
-            int n_dims = ASRUtils::extract_dimensions_from_ttype(array_type, m_dims);
-            LCOMPILERS_ASSERT(n_dims == value_rank);
+            LCOMPILERS_ASSERT(ASRUtils::extract_dimensions_from_ttype(array_type, m_dims) == value_rank);
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:4656]: This change introduces a bug in Release mode, fixed in https://github.com/lfortran/lfortran/pull/1771.

**Pull request merged**